docs/content/sips/016-inbound-websockets.md

```diff
@@ -0,0 +1,189 @@
+    // failure. Applications which require delivery confirmation
+    // must handle that themselves.
+    get: func() -> option<result<result<_, error-code>>>;
+}
+
+interface inbound-websocket-receive {
+  use wasi:http/types.{incoming-request, response-outparam};
+  use websocket-types.{frame, close-frame};
+
+  // Construct a new handler for the specified request, which the
+  // client has requested to be upgraded to a WebSocket.
+  //
+  // - `id`: a globally unique identifier which may be used to send
+  //   frames to this WebSocket using
+  //   `outgoing-websocket#send-frame`. Note that any application
+  //   may use this `id` to send frames to the WebSocket for as
+  //   long as it is connected, which means it should be treated as
+  //   a secret and shared only with trusted parties.
+  //
+  // - `request`: a resource providing the request's method, path,
+  //   headers, etc.
+  //
+  // - `response-out`: a resource which may be used to send a
+  //   response to the client. If (and only if) the response has a
+  //   status code of 101, then the connection will be upgraded to a
+  //   WebSocket.
+  handle-new: func(id: string, request: incoming-request, response-out: response-outparam);
+
+  // Handle the most recent frames received for the specified
+  // WebSockets. `frames` consists of a mapping from WebSocket ids
+  // to ordered lists of frames.
+  handle-frames: func(frames: list<tuple<string, list<received-frame>>);
+}
+
+interface inbound-websocket-send {
+  use websocket-types.{frame, future-send-frame};
+
+  // Attempt to send the specified frame to specified WebSocket.
+  //
+  // This will fail if the WebSocket is no longer open. Otherwise,
+  // if the frame is an instance of `close`, the connection will
+  // be closed after sending the frame.
+  send-frame: func(id: string, frame: frame) -> future-send-frame;
+}
+
+world inbound-websocket-handler {
+  // Incoming requests which do not request a WebSocket upgrade
+  // will be passed to `wasi:http/incoming-handler#handle`
+  include wasi:http/proxy;
+
+  // Incoming requests which request a WebSocket upgrade will be
+  // passed to `incoming-websocket-receive#handle-new`.
+  export inbound-websocket-receive;
+
+  import inbound-websocket-send;
+}
+```
+
+### Spin Implementation
+
+We’ll add a new trigger type called `websocket` which will be handled by the `spin-trigger-http` crate. This crate will handle normal (non-WebSocket) incoming requests as usual: instantiating the component, calling the `wasi:http/incoming-handler#handle` method, and discarding the instantiation when the entire response has been sent. For WebSocket requests (i.e. those with headers requesting a connection upgrade per the WebSocket protocol), it will instead call `fermyon:spin/inbound-websocket-receive#handle-new`, and if that produces a valid 101 ("Switching Protocols") response, the host component will discard the instance and upgrade the connection to a WebSocket.
```

Comment:
Do we need a new trigger type? The http trigger already supports multiple worlds (spin, wasi, wagi) chosen at runtime based on the shape of the component. Any inherent reason why it couldn't also choose based on the shape of the request?

Alternatively, maybe this could _extend_ the WASI http inbound world with this `handle-frames` export and a new e.g. `upgrade-websocket: function(inbound-request) -> token`.

Reply:
To be clear, my thinking was that the same code would be shared between `http` and `websocket` trigger types.  And yes, extending the wasi-http world is exactly what I'm proposing -- note the `include wasi:http/proxy;` line in the world.

The reason I'm proposing a new trigger type named `websocket` is to make it clear to Spin that it can rely on the component to export `inbound-websocket-receive`.  It's not necessary, since Spin can always ask the component what it exports, but I figured it's best to be explicit -- the same way we're explicit about other trigger types even though we could arguably guess based on the exports.